File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if(n == 1) {
            return x;
        }
        if(n == 0) {
            return 1;
        }
        if(n == -1) {
            return 1/x;
        }
        if(x == 0) {
            return 0;
        }

        double res = myPow(x, n/2);

        if(n % 2 == 0) {
            return res * res;
        } else {
            if(n < 1) {
                return 1/x * res * res;
            }
            return x * res * res;
        }
    }
};

File: 0050-powx-n/0050-powx-n.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
    if(n === 0) {
        return 1;
    }
    if(n === -1) {
        return 1/x;
    }

    let res = myPow(x, Math.floor(n/2));

    if(n % 2 === 0) {
        return res * res;
    } else {
        if(n < 0) {
            return 1/x * res * res;
        } else {
            return x * res * res;
        }
    }
};